src/main/java/de/tum/cit/aet/artemis/programming/service/AutomaticProgrammingExerciseCleanupService.java

```diff
@@ -82,49 +89,66 @@ public void cleanup() {
-        programmingExercises.addAll(programmingExerciseRepository.findAllByRecentExamEndDate(endDate1, endDate2));
+        log.info("Search for exercises with course or exam date from {} until {}", earliestDate, latestDate);
+        var programmingExercises = programmingExerciseRepository.findAllByRecentCourseEndDate(earliestDate, latestDate);
+        programmingExercises.addAll(programmingExerciseRepository.findAllByRecentExamEndDate(earliestDate, latestDate));
         log.info("Found {} programming exercise to clean local template, test and solution: {}", programmingExercises.size(),
                 programmingExercises.stream().map(ProgrammingExercise::getProjectKey).collect(Collectors.joining(", ")));
-        for (var programmingExercise : programmingExercises) {
-            gitService.deleteLocalRepository(programmingExercise.getVcsTemplateRepositoryUri());
-            gitService.deleteLocalRepository(programmingExercise.getVcsSolutionRepositoryUri());
-            gitService.deleteLocalRepository(programmingExercise.getVcsTestRepositoryUri());
-            gitService.deleteLocalProgrammingExerciseReposFolder(programmingExercise);
+        if (!programmingExercises.isEmpty()) {
+            for (var programmingExercise : programmingExercises) {
+                gitService.deleteLocalRepository(programmingExercise.getVcsTemplateRepositoryUri());
+                gitService.deleteLocalRepository(programmingExercise.getVcsSolutionRepositoryUri());
+                gitService.deleteLocalRepository(programmingExercise.getVcsTestRepositoryUri());
+                gitService.deleteLocalProgrammingExerciseReposFolder(programmingExercise);
+            }
+            log.info("Finished cleaning local template, test and solution repositories");
+        }
+    }
+
+    private void cleanStudentParticipationsRepositories(ZonedDateTime earliestDate, ZonedDateTime latestDate) {
+        log.info("Search for exercises with due date from {} until {}", earliestDate, latestDate);
+        // Get all relevant participation ids
+        Pageable pageable = Pageable.ofSize(STUDENT_PARTICIPATION_CLEANUP_BATCH_SIZE);
+        Page<String> uriBatch = programmingExerciseStudentParticipationRepository.findRepositoryUrisForGitCleanupByRecentDueDateOrRecentExamEndDate(earliestDate, latestDate,
+                pageable);
+        log.info("Found {} student participations to clean local student repositories in {} batches.", uriBatch.getTotalElements(), uriBatch.getTotalPages());
+        if (uriBatch.getTotalElements() > 0) {
+            var batchStream = Stream.iterate(uriBatch, page -> page.getNumber() <= page.getTotalPages(), page -> programmingExerciseStudentParticipationRepository
+                    .findRepositoryUrisForGitCleanupByRecentDueDateOrRecentExamEndDate(earliestDate, latestDate, page.nextPageable()));
+            batchStream.forEach(page -> page.forEach(this::deleteLocalRepositoryByUriString));
```

Comment:
On the last page `page.nextPageable()` wraps over to `Pageable.unpaged()` which always fulfills the loop condition.
`Stream.iterate()` might not be a useful abstraction here. Try this:
```suggestion
            while (true) {
                uriBatch.forEach(this::deleteLocalRepositoryByUriString);
                if (!uriBatch.hasNext()) {
                    break;
                }
                uriBatch = programmingExerciseStudentParticipationRepository.findRepositoryUrisForGitCleanupByRecentDueDateOrRecentExamEndDate(earliestDate, latestDate, uriBatch.nextPageable());
            }
```

Reply:
Thanks a lot. I'll check this in the next few days for other alternatives. Stephan doesn't like do while, I think while true is not the way to go and streams aren't good either in this situation. We'll see.

Reply:
We can extract part of the first loop iteration to get an explicit loop condition:
```suggestion
            uriBatch.forEach(this::deleteLocalRepositoryByUriString);
            while (uriBatch.hasNext()) {
                uriBatch = programmingExerciseStudentParticipationRepository.findRepositoryUrisForGitCleanupByRecentDueDateOrRecentExamEndDate(earliestDate, latestDate, uriBatch.nextPageable());
                uriBatch.forEach(this::deleteLocalRepositoryByUriString);
            }
```

This, however, is not strictly DRY.

Reply:
I just played around it myself and actually came to the exact same conclusion as you without having it in mind.
So I think it's the best we can do. Also, I don't mind it in terms of DRY - I already extracted the deletion functionality, for the method call the parameters are different. So we have one one-liner which repeats itself once. That's fine.